src/session.rs

```diff
@@ -147,6 +156,82 @@ impl NgrokSessionBuilder {
+    /// Roughly corresponds to the [root_cas parameter in the ngrok docs].
+    ///
+    /// [root_cas parameter in the ngrok docs]: https://ngrok.com/docs/ngrok-agent/config#root_cas
+    #[napi]
+    pub fn tls_config(&mut self, cert_bytes: Uint8Array) -> &Self {
+        let mut root_store = rustls::RootCertStore::empty();
+        let mut cert_pem = io::Cursor::new(cert_bytes);
+        root_store.add_parsable_certificates(
+            rustls_pemfile::read_all(&mut cert_pem)
+                .expect("a valid root certificate")
+                .into_iter()
+                .filter_map(|it| match it {
+                    Item::X509Certificate(bs) => Some(bs),
+                    _ => None,
+                })
+                .collect::<Vec<_>>()
+                .as_slice(),
+        );
+
+        let tls_config = rustls::ClientConfig::builder()
+            .with_safe_defaults()
+            .with_root_certificates(root_store)
+            .with_no_client_auth();
+
+        let mut builder = self.raw_builder.lock();
+        *builder = builder.clone().tls_config(tls_config);
+        self
+    }
+
+    /// Configures a function which is called to establish the connection to the
+    /// ngrok service. Use this option if you need to connect through an outbound
+    /// proxy. In the event of network disruptions, it will be called each time
+    /// the session reconnects. If the handler responds with a string it will be
+    /// used as the new address to connect to, e.g. "192.168.1.1:443".
+    #[napi(ts_args_type = "handler: (addr: string, error?: string) => string")]
+    pub fn connector(&mut self, env: Env, handler: JsFunction) -> &Self {
+        // create threadsafe function
+        let tsfn = create_tsfn(env, handler);
+
+        // register connect handler. this needs the return value, so cannot use call_tsfn().
+        let mut builder = self.raw_builder.lock();
+        *builder = builder.clone().connector(
+            move |addr: String, tls_config: Arc<ClientConfig>, err: Option<AcceptError>| {
+                let tsfn = tsfn.clone();
+                let mut args = vec![addr.clone()];
+                if err.is_some() {
+                    args.push(err.clone().unwrap().to_string());
+                }
+
+                async move {
+                    // call javascript handler
+                    let result: Option<String> = tsfn
+                        .clone()
+                        .lock()
+                        .await
+                        .call_async(args)
+                        .await
+                        .map_err(|_e| ConnectError::Canceled)?;
+                    // call the upstream connector
+                    let new_addr = match result {
+                        Some(result) => result,
+                        None => addr,
+                    };
+                    default_connect(new_addr, tls_config, err).await
+                }
+            },
+        );
+        self
+    }
```

Comment:
Yeah, there are limits in what is exposed in the client language here, maybe I should have done this as a Draft :) Wanted this change to be able to set the TLS certificate to use, and the address to connect to, so if a preconfigured forwarding proxy is set up it could be used.

This has a secondary effect of getting a callback on connects and reconnects, which is likely useful on its own for for the user handling in some way.

There's a bunch of copied code for the cert setting, but wanted to see what you'd think about that being the only modifiable piece, since on the rust side you can change dozens of settings in there. If just setting the cert is useful I could see making a convenience function for that on the rust side for all the sdk's to use.

Given what you're saying, maybe SessionBuilder exposes something like:
// bare minimum to hit an nginx proxy_pass or something, can iteratively add more pieces to it.
`connect_settings(address, cert)`
// a simple callback when a connection is about to happen to notify the user.
`handle_connection(handle)`

Reply:
> There's a bunch of copied code for the cert setting, but wanted to see what you'd think about that being the only modifiable piece, since on the rust side you can change dozens of settings in there. If just setting the cert is useful I could see making a convenience function for that on the rust side for all the sdk's to use.

Yeah, probably worth it. We have that on the Go side - I just omitted it in Rust because I'm lazy and being able to set the whole TLSConfig lets you do that with some extra hoops :stuck_out_tongue: 

> This has a secondary effect of getting a callback on connects and reconnects, which is likely useful on its own for the user handling in some way.

Maybe we make it similar to Go, who also doesn't get to mess with the actual connection? There, we have a connect handler and a disconnect handler. The disconnect handler gets the error, and the connect handler just gets... the session? Not really sure how useful they are though.

> Given what you're saying, maybe SessionBuilder exposes something like:
> // bare minimum to hit an nginx proxy_pass or something, can iteratively add more pieces to it.
> connect_settings(address, cert)

Isn't this just:
```
  .server_addr(address)
  .ca_cert(cert)
```

Reply:
> Isn't this just:
>  .server_addr(address)
>  .ca_cert(cert)

ah, but that would be consistent!

Ok, I'll work on updating this PR with those changes, and see about some cert convenience on the rust side.

Reply:
Latest update has `ca_cert`, `handle_connection` and `handle_disconnection`. Plan to move ca_cert internals upstream in a later change.